File: lib/dijkstra.py
graph = {}
graph['start'] = {}
graph['start']['a'] = 6
graph['start']['b'] = 2

graph['a'] = {}
graph['a']['end'] = 1

graph['b'] = {}
graph['b']['a'] = 3
graph['b']['end'] = 5

graph['end'] = {}

infinity = float('inf')
costs = {}
costs['a'] = 6
costs['b'] = 2
costs['end'] = infinity

parents = {}
parents['a'] = 'start'
parents['b'] = 'start'
parents['end'] = None

processed = []

def findLowestCostNode(costs):
  lowestCost = float('inf')
  lowestCostNode = None
  for node in costs:
    cost = costs[node]
    if cost < lowestCost and node not in processed:
      lowestCost = cost
      lowestCostNode = node
  return lowestCostNode
  
node = findLowestCostNode(costs)
while node is not None:
  cost = costs[node]
  neighbors = graph[node]
  for n in neighbors.keys():
    newCost = cost + neighbors[n]
    if costs[n] > newCost:
      costs[n] = newCost
      parents[n] = node
  processed.append(node)
  node = findLowestCostNode(costs)

File: lib/dijkstra.js
const findLowestCostNode = (costs, processed) => {
  return Object.keys(costs).reduce((lowest, node) => {
    if(lowest === null || costs[node] < costs[lowest]) {
      if(!processed.includes(node)) {
        lowest = node;
      }
    }
    return lowest
  }, null);
};

module.exports = dijkstra = graph => {
  
  const costs = Object.assign({ finish: Infinity }, graph.start);

  const parents = { finish: null };
  for(let i in graph.start) {
    parents[i] = 'start';
  }

  const processed = [];

  let node = findLowestCostNode(costs, processed);

  while(node) {
    let cost = costs[node];
    let children = graph[node];
    for(let j in children) {
      let newCost = cost + children[j];
      if(!costs[j]) {
        costs[j] = newCost;
        parents[j] = node;
      }
      if(costs[j] > newCost) {
        costs[j] = newCost;
        parents[j] = node;
      }
    }
    processed.push(node);
    node = findLowestCostNode(costs, processed);
  }

  let optimalPath = ['finish'];
  let parent = parents.finish;
  while(parent) {
    optimalPath.push(parent);
    parent = parents[parent];
  }
  optimalPath.reverse();

  const results = {
    distance: costs.finish,
    path: optimalPath
  };

  return results;
};

